File: hella/src/main/java/com/lydiaschiff/hella/renderer/GreyscaleRsRenderer.java
package com.lydiaschiff.hella.renderer;

import android.os.Build;
import android.renderscript.Allocation;
import android.renderscript.Element;
import android.renderscript.RenderScript;
import android.renderscript.ScriptIntrinsicColorMatrix;
import android.support.annotation.RequiresApi;

import com.lydiaschiff.hella.RsRenderer;

/**
 * Render an image allocation in greyscale.
 */
@RequiresApi(17)
public class GreyscaleRsRenderer implements RsRenderer {

    private ScriptIntrinsicColorMatrix greyscaleScript;

    @Override
    public void renderFrame(RenderScript rs, Allocation in, Allocation out) {
        if (greyscaleScript == null) {
            greyscaleScript = createScript(rs, in.getElement());
            greyscaleScript.setGreyscale();
        }
        greyscaleScript.forEach(in, out);
    }

    @Override
    public String getName() {
        return "Greyscale with ScriptIntrinsicColorMatrix";
    }

    @Override
    public boolean canRenderInPlace() {
        return true;
    }

    private static ScriptIntrinsicColorMatrix createScript(RenderScript rs, Element e) {
        if (Build.VERSION.SDK_INT >= 19) {
            return ScriptIntrinsicColorMatrix.create(rs);
        } else {
            return ScriptIntrinsicColorMatrix.create(rs, e);
        }
    }
}

File: hella/src/main/java/com/lydiaschiff/hella/renderer/GreyscaleRsRenderer.kt
package com.lydiaschiff.hella.renderer

import android.os.Build
import android.renderscript.Allocation
import android.renderscript.Element
import android.renderscript.RenderScript
import android.renderscript.ScriptIntrinsicColorMatrix
import androidx.annotation.RequiresApi
import com.lydiaschiff.hella.RsRenderer

/**
 * Render an image allocation in greyscale.
 */
@RequiresApi(17)
class GreyscaleRsRenderer : RsRenderer {

    private var greyscaleScript: ScriptIntrinsicColorMatrix? = null

    override fun renderFrame(rs: RenderScript, inAlloc: Allocation, outAlloc: Allocation) {
        val script = greyscaleScript ?: createScript(rs, inAlloc.element).also {
            it.setGreyscale()
            greyscaleScript = it
        }
        script.forEach(inAlloc, outAlloc)
    }

    override val name = "Greyscale with ScriptIntrinsicColorMatrix"

    override val canRenderInPlace = true

    companion object {
        private fun createScript(rs: RenderScript, e: Element): ScriptIntrinsicColorMatrix {
            return if (Build.VERSION.SDK_INT >= 19) {
                ScriptIntrinsicColorMatrix.create(rs)
            } else {
                ScriptIntrinsicColorMatrix.create(rs, e)
            }
        }
    }
}